swiftui/ChartboostMediationDemo/Ads/AdAnchor.swift

```diff
@@ -0,0 +1,57 @@
+// Copyright 2022-2023 Chartboost, Inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file.
+
+//
+//  AdAnchor.swift
+//  ChartboostMediationDemo
+//
+//  Copyright © 2023 Chartboost. All rights reserved.
+//
+
+import SwiftUI
+
+extension View {
+
+    /// The Chartboost Mediation SDK requires a `UIViewController` to be provided in its full screen show and banner load APIs.
+    /// The purpose of this `AdAnchor` is to be able to provide a `UIViewController` within the scope of a SwiftUI `View` that
+    /// will be the host of advertisements.  For any such `View`, add this view modifier to a view within it.  When the view is
+    /// presented, the `didMake` callback will provide the instance of a `UIViewController` that can be used to
+    /// provide to the Chartboost Mediation SDK APIs,
+    func adAnchor(didMake: @escaping (UIViewController) -> ()) -> some View {
+        modifier(AdAnchorViewModifier(didMake: didMake))
+    }
+}
+
+private struct AdAnchor: UIViewControllerRepresentable {
+    typealias UIViewControllerType = UIViewController
+
+    let didMake: (UIViewController) -> ()
+
+    func makeUIViewController(context: Context) -> UIViewController {
+        let viewController = UIViewController()
```

Comment:
I'm not sure this is the best way to provide the view controller needed to show ads.
If I'm not wrong we are never presenting this view controller on screen, which means it is a detached vc. Presenting vcs on top of detached vcs may lead to errors. I remember getting some console warnings about this in the past, but correct me if my knowledge is outdated.

Maybe we could use the top-most view controller instead? We could have the InterstitialAdController/RewardedAdController/BannerAdController class obtain it so SwiftUI views don't need to care about vcs or anchors.
This is how it was solved in the SwiftUI demo app the ads sdk team built. This is a snippet I stole from that codebase:
```
extension UIApplication {
    class func topViewController(base: [UIWindow] = UIApplication.shared.windows) -> UIViewController? {
        let keyWindow = base.filter {$0.isKeyWindow}.first
        if var topViewController = keyWindow?.rootViewController {
            while let presentedViewController = topViewController.presentedViewController {
                topViewController = presentedViewController
            }
            return topViewController
        }
        return nil
    }
}
```

Reply:
Cool, I'll give it a try.

Reply:
@daniel-barros Question.  If you feel that using that extension is the way to go, then why doesn't the Chartboost Mediation SDK just do something like that internally and not require view controllers to be passed into the APIs at all?

Reply:
This is more modern:

```
    class var topViewController: UIViewController? {
        guard
            let windowScene = UIApplication.shared.connectedScenes.compactMap({ $0 as? UIWindowScene }).first,
            let keyWindow = windowScene.keyWindow,
            var rootViewController = keyWindow.rootViewController
        else {
            return nil
        }
        while let presentedViewController = rootViewController.presentedViewController {
            rootViewController = presentedViewController
        }
        return rootViewController
    }
```

Reply:
So this solution is not perfect either for UIKit apps. Some sophisticated users may have multiple windows (if I remember right you can windows inside windows although I don't know why you'd do that), also there can be multiple scenes... So it's safer to let pubs define the VC themselves, this is what most SDKs do.

Feel free to do more research on this issue. My suggestion is likely not the definitive solution. Or we can create a ticket to do the research.

Reply:
Well, we know of 2 ways for sure (my original way and this other way), both of which would work fine for this demo.  Do we really need to know more?

Reply:
Yeah I think this is good enough for now given that SwiftUI pubs are not a sizable part of our user base.